File: python/152_maximum_product_subarray.py
#!/usr/bin/env python3

import sys


class Solution(object):
    def maxProductWithoutZeros(self, nums):
        if not nums:
            return - sys.maxsize - 1
        if len(nums) == 1:
            return nums[0]
        l = 1
        m = 1
        r = 1
        i = 0
        j = len(nums) - 1
        while i < j:
            if nums[i] < 0:
                break
            l *= nums[i]
            i += 1
        while i < j:
            if nums[j] < 0:
                break
            r *= nums[j]
            j -= 1
        for k in range(i + 1, j):
            m *= nums[k]

        if i == j:
            if nums[i] > 0:
                return l * nums[i]
            return max(l, r)

        l *= nums[i]
        r *= nums[j]
        if m > 0:
            return l * m * r
        else:
            return m * min(l, r)

    def maxProduct(self, nums):
        if not nums:
            return 0
        j = 0
        product = - sys.maxsize - 1
        for i in range(len(nums)):
            if nums[i] == 0:
                product = max(product, self.maxProductWithoutZeros(nums[j:i]))
                j = i + 1
        if j != 0:
            product = max(product, 0)
        product = max(product, self.maxProductWithoutZeros(nums[j:]))
        return product


if __name__ == '__main__':
    solution = Solution()
    print(solution.maxProduct([2, 3, -2, 4]), 6)
    print(solution.maxProduct([2, 3, -2]), 6)
    print(solution.maxProduct([-2, 4]), 4)
    print(solution.maxProduct([2, 4]), 8)
    print(solution.maxProduct([2, 3, 2, 4]), 48)
    print(solution.maxProduct([2, -1, 3, -2, 5, -3, 4]), 360)
    print(solution.maxProduct([-2, 0, -1]), 0)
    print(solution.maxProduct([0]), 0)
    print(solution.maxProduct([6, 3, -10, 0, 2]), 18)

File: c++/152-maximum-product-subarray.cpp
#include <gtest/gtest.h>
#include <iostream>

using namespace std;

class Solution {
   public:
    int maxProduct(vector<int>& nums) {
        int m = numeric_limits<int>::min();
        int n = numeric_limits<int>::max();
        int p = 1;
        for (int i = 0; i < nums.size(); i++) {
            p *= nums[i];
            if (p != 0) {
                m = max(m, p);
                n = min(n, p);
            } else {
                m = max(m, 0);
                p = 1;
            }
        }
        p = 1;
        for (int i = nums.size() - 1; i >= 0; i--) {
            p *= nums[i];
            if (p != 0) {
                m = max(m, p);
                n = min(n, p);
            } else {
                m = max(m, 0);
                p = 1;
            }
        }

        return m;
    }
};

int main(int argc, char* argv[]) {
    testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}

TEST(test, case1) {
    Solution solution;
}